universal-application-tool-0.0.1/app/models/Question.java

```diff
@@ -1,34 +1,77 @@
 package models;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableMap;
 import io.ebean.annotation.DbJsonB;
-import io.ebean.text.json.EJson;
-import java.io.IOException;
-import java.util.Map;
+import java.util.Locale;
+import java.util.Optional;
 import javax.persistence.Entity;
+import javax.persistence.PostLoad;
+import javax.persistence.PostPersist;
+import javax.persistence.PostUpdate;
+import javax.persistence.PreUpdate;
 import javax.persistence.Table;
 import play.data.validation.Constraints;
+import services.question.QuestionDefinition;
 
 @Entity
 @Table(name = "questions")
 public class Question extends BaseModel {
-  private static final long serialVersionUID = 1L;
 
-  public Map<String, Object> getObject() {
-    return object;
+  private QuestionDefinition questionDefinition;
+
+  private @Constraints.Required long version;
+
+  private @Constraints.Required String path;
+
+  private @Constraints.Required String name;
+
+  private @Constraints.Required String description;
+
+  private @Constraints.Required @DbJsonB ImmutableMap<Locale, String> questionText;
+
+  private @DbJsonB ImmutableMap<Locale, String> questionHelpText;
+
+  public Question(QuestionDefinition questionDefinition) {
+    this.questionDefinition = checkNotNull(questionDefinition);
+    this.version = questionDefinition.getVersion();
+    this.path = questionDefinition.getPath();
+    this.name = questionDefinition.getName();
+    this.description = questionDefinition.getDescription();
+    this.questionText = questionDefinition.getQuestionText();
+    this.questionHelpText = questionDefinition.getQuestionHelpText().orElse(null);
   }
 
-  public void setObject(Map<String, Object> object) {
-    this.object = object;
+  /** Populates column values from {@link QuestionDefinition}. */
+  @PreUpdate
+  public void persistChangesToQuestionDefinition() {
+    this.id = questionDefinition.getId();
+    this.version = questionDefinition.getVersion();
+    this.path = questionDefinition.getPath();
+    this.name = questionDefinition.getName();
+    this.description = questionDefinition.getDescription();
+    this.questionText = questionDefinition.getQuestionText();
+    this.questionHelpText = questionDefinition.getQuestionHelpText().orElse(null);
```

Comment:
If this was `orElse(ImmutableMap.of())`, would `loadQuestionDefinition` be able to use `Optional.of(this.questionHelpText)` and we could avoid nulls altogether?

Reply:
Agree that anytime we have null values is cause for concern, but I think we want nulls here - if a question doesn't have help text, we want `Optional.empty()`, not an `Optional` with an empty map